### ISSUE TITLE
Command-line interface

### ISSUE THREAD

**Issue by mikeday** [yeslogic/fathom#227]
Some thoughts for a future `fathom-cli` package that uses the DDL interpreter to provide what is essentially a supercharged version of the `file` command and/or the equivalent of `grep` for binary files. :nerd_face:

## Syntax
```
fathom
  [--format=NAME | --format-file=FILE]
  [--info | --check | --dump | --query=QUERY | --query-file=FILE ]
  FILES...
```

## Format Options

**--format**=*NAME* is an optional argument for specifying the name of the format description in the installed catalog, for example "opentype", that will be used to process the files.

**--format-file**=*FILE* is an optional argument for specifying the file containing the format description, for example "path/to/myformat.ddl", that will be used to process the files.

If neither format option is provided the files will be processed after best-effort identification using the installed catalog of format descriptions.

## Command Options

**--info** prints a brief summary of each file after processing with the appropriate format description. This may include printing the type and version and any important details about the file.

**--check** attempts to validate each file against the format description and prints a report of any discrepancies.

**--dump** prints the complete contents of each file as described by the format description. (This probably implies validation as well?)

**--query**=*QUERY*  and **--query-file**=*FILE* take a format query and apply it to each file, printing the results. (Haven't even begun to describe format queries but the idea would be to dump only values of specified types that satisfy certain constraints, for example "Ligature where componentCount > 2").

If no command option is specified it defaults to `--info`.

## Examples
```
fathom FILES...
```
Print a summary of each file after best-effort identification using the installed catalog of format descriptions.

```
fathom --ddl FILE.DDL FILES...
```
Print a summary of each file after applying the specified format description.

```
fathom --dump FILES...
```
Print a complete dump of the contents of each file.

**Comment by toothbrush** [comment]
This looks good! I think it might be an interesting "learn yourself a Rust" project for me to try to have a stab at making a CLI for this.

Would we give the user a warning if they specified both `--summary` and `--complete`?  To me that seems like a contradiction in terms (partly i'm asking to try and grow my understanding of this ecosystem-to-be :smiley_cat:).

**Comment by brendanzab** [comment]
Yeah, seems like a good first project! [structopt](https://lib.rs/crates/structopt) is a pretty nice way for declaring these CLIs. I use this in [pikelet-cli](https://github.com/pikelet-lang/pikelet/tree/next/pikelet-cli).

**Comment by wezm** [comment]
There's an official Rust CLI working group too. They have an intro book: https://rust-cli.github.io/book/index.html

**Comment by mikeday** [comment]
> Would we give the user a warning if they specified both `--summary` and `--complete`?

Yes that would be contradictory (although now that I test it I note that grep allows you to specify -l and -h without complaint, with -l winning).

Also I'm tempted to switch `--complete`, which is an adjective, back to `--dump`, which is a noun as well as a verb!

**Comment by toothbrush** [comment]
On 2020-09-10 at 12:16 AEST+1000, quoth Michael Day 
<notifications@github.com>:
>> Would we give the user a warning if they specified both 
>> `--summary` and `--complete`?
>
> Yes that would be contradictory (although now that I test it I 
> note that grep allows you to specify -l and -h without 
> complaint, with -l winning).

Fair enough!  I think `grep` is a better example than `git` as far 
as UIs go, in any case.

> Also I'm tempted to switch `--complete`, which is an adjective, 
> back to `--dump`, which is a noun as well as a verb!

To be fair, `--complete` is a verb too. 😛  If i'm being 
facetious.  But i agree about `--dump` being clearer / preferable.
